### PR TITLE
Fix panic when enable titan

### PR DESCRIPTION
**Pull request by ethercflow**
Issue Number: close tikv/tikv#13038

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

### What is changed and how it works?

Tests 

- Manual test 
  - sysbench prepare 

Side effects

- Performance regression
    - None
- Breaking backward compatibility
   - None 

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by tabokie — APPROVED** [review]

**Comment by tabokie** [comment]
Please fix test in test_titan to use multi batch write.

**Comment by ethercflow** [comment]
> Please fix test in test_titan to use multi batch write.

Sure

**Comment by ethercflow** [comment]
> Please fix test in test_titan to use multi batch write.

@tabokie I suddenly realized that multi batch write is enabled by default, and I make sure it at IDC without this PR.

**Review by Connor1996 — COMMENTED** [review]
Need a case to cover titan write

**Comment by ethercflow** [comment]
> Need a case to cover titan write. Maybe enable titan in `test_raftstore` integration tests

@Connor1996 @tabokie Add a write wb test in `test_titan`, as only [more than one wb](https://github.com/tikv/tikv/pull/12899/files#diff-3be953d50746ae8872ed880f72d3c7226cda7e6470b82b9a5e77bfd2e337711dR101 )  will call `multi_batch_write` directly,  otherwise it will call `write_opt`, that's why existing titan tests are not painic with `enable_multi_write_batch` enabled.

**Review by tabokie — COMMENTED** [review]
Why not change the `index>0` to use `support_multi_batch_write`? And eventually we can remove the flag in RocksDB.

**Comment by ethercflow** [comment]
> Why not change the `index>0` to use `support_multi_batch_write`? And eventually we can remove the flag in RocksDB.

Because `write_opt` receives `WriteBatch` as a parameter, and `multi_write_batch` receives an `vec`, I think it is more straightforward to judge which interface to use by `index`. (As when we call `db.put`, `db.put` will also call `write` and use `flag` to judge which `write impl` to use).

If use `support_multi_batch_write`, the as_inner need changed to:

``` rust
pub fn as_inner(&self) -> &[RawWriteBatch] {
    if self.index > 0 {
        &self.wbs[0..self.index]
    } else {
         &[self.wbs[0]]
    }
}
```

Write once, judge twice

**Comment by tabokie** [comment]
@ethercflow It's always `&self.wbs[0..=self.index]`. No need to add an if.

**Comment by ethercflow** [comment]
> @ethercflow It's always `&self.wbs[0..=self.index]`. No need to add an if.

@tabokie `&self.wbs[0..0]` will return an empty `[]`.

**Comment by tabokie** [comment]
@ethercflow It's `&self.wbs[0..=0]`, please check.

**Comment by ethercflow** [comment]
> @ethercflow It's `&self.wbs[0..=0]`, please check.

Oh, you're right, I left `=` :(

File: tests/integrations/storage/test_titan.rs [hunk 1]
```diff
@@ -431,3 +431,59 @@ fn test_delete_files_in_range_for_titan() {
         Some((Key::from_raw(b"b"), b"b_value".to_vec())),
     );
 }
+
+#[test]
+fn test_writebatch_for_titan() {
```
Comment (Connor1996) [inline tests/integrations/storage/test_titan.rs:436]: it easily misses the new API after changes in raftstore. Better just choose some basic tests under `tests/integrations/raftstore` with titan enabled

Comment (ethercflow) [reply tests/integrations/storage/test_titan.rs:436]: Got it, I'll choose some one, thanks!

**Comment by ethercflow** [comment]
> Why not change the `index>0` to use `support_multi_batch_write`? And eventually we can remove the flag in RocksDB.

@tabokie changed to `support_write_batch_vec`, PTAL again, thanks!

File: tests/integrations/storage/test_titan.rs (hunk 1 above)
Comment (ethercflow) [reply tests/integrations/storage/test_titan.rs]: > it easily misses the new API after changes in raftstore. Better just choose some basic tests under `tests/integrations/raftstore` with titan enabled

@Connor1996 I chose a few tests with `titan` enabled  include `put`, `delete`,`snap` interface. PTAL if it's enough to cover, thanks!

**Pull request merged**